File: all-submissions/1092-maximum-difference-between-node-and-ancestor/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxValue = 0;
    int minValue = 0;

    public int helper(TreeNode root, int maxValue, int minValue){
        if (root == null) return maxValue-minValue;
        maxValue = Math.max(root.val, maxValue);
        minValue = Math.min(root.val, minValue);
        int left = helper(root.left, maxValue, minValue);
        int right = helper(root.right, maxValue, minValue);
        return Math.max(left, right);
    }

    public int maxAncestorDiff(TreeNode root) {
        if (root == null) return 0;
        maxValue = root.val;
        minValue = root.val;
        return helper(root, maxValue, minValue);
                
    }
}

File: all-submissions/1092-maximum-difference-between-node-and-ancestor/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        def dfs_helper(node, curr_max, curr_min):
            if node is None:
                return curr_max - curr_min
            curr_max = max(node.val, curr_max)
            curr_min = min(node.val, curr_min)
                    
            left = dfs_helper(node.left, curr_max, curr_min)
            right = dfs_helper(node.right, curr_max, curr_min)
            return max(left, right)
        return dfs_helper(root, root.val, root.val)
